### ISSUE TITLE
Support hair & fur export

### ISSUE THREAD

**Issue by kutomer** [BabylonJS/Exporters#51]
Hello :)
I think it will be very useful if we will be able to export [3dsmax hair & fur modifer](https://knowledge.autodesk.com/support/3ds-max/learn-explore/caas/CloudHelp/cloudhelp/2016/ENU/3DSMax/files/GUID-4BCFC5E4-266F-43AA-A552-77CC241F9D1E-htm.html) as [Babylon.FurMaterial](http://doc.babylonjs.com/extensions/fur)

in the meantime do we have any other way to export fur & hair from 3dsmax to Babylon.js?

Thanks!

**Comment by deltakosh** [comment]
This is a good idea. So far the only way would be to do it at load time in JS :(

**Comment by kutomer** [comment]
Thanks a lot for the quick reply!
Do you think that it's something can be added to a backlog or something?
Would love to help to develop this feature but guidance is needed.

**Comment by deltakosh** [comment]
Obviously :)
You can perhaps familiarize yourself with the C# code of the exporter. Then we can discuss where in the json file we can export it (it is not a big deal as we can export all kind of materials)

**Comment by kutomer** [comment]
Do you have any documentation for the exporter which I can begin with?

**Comment by deltakosh** [comment]
I would say tags just loadibg the VS project should be a good start
Project is pretty straightforward

**Comment by kutomer** [comment]
Thanks,

1. I couldn't find a way to debug the plugin (read that I need to be a certified Autodesk developer), the way I'm doing it now is with `RaiseMessage` and for that, I need to rebuild and relaunch 3dsmax what makes the development process much slower.
2. I guess I'm doing something wrong but `meshMode.IGameObject.NumModifiers.ToString()` returns 0 for me, have any idea?

**Comment by noalak** [comment]
Hi kutomer,

1) Doing the same
2) Hair and Fur modifier is a [World Space Modifier](http://help.autodesk.com/view/3DSMAX/2017/ENU/?guid=__files_GUID_DA7227A8_576A_4D3D_9254_4DFA1C49B08E_htm). Don't really know much about it but [find attached the code](https://github.com/BabylonJS/Exporters/files/1611117/hairAndFur_code.txt) you are probably looking for.

With that you get all parameters from Hair and Fur modifier. Seems like there is no interface provided by the API, so you need to retreive each parameter by id. You need to adapt the getter method (GetFloat, GetColor...) to the parameter type of each parameter, otherwise it would return the default value (0 for a float).

Be aware that an object might have several WSMs. To retreive the Hair and Fur one you can check its name but this process is ugly and certainly language sensitive (i experimented this for materials' type). If you go for that, tag it with a TODO :)

**Comment by kutomer** [comment]
@noalak awesome, thanks, I'll do that.

1. So is that fine if I'll support just English for now and add a TODO like: 
`materialNode.MaterialClass.ToLower() == "physisches material" || // German // TODO - check if translation is ok` (`Exporter/BabylonExporter.Material.cs`)
(just wanna make sure you guys will approve my PR)

2. After exporting all the data from the Modifier I want to create a new Material in the exported babylon file`materials` array with customType `FurMaterial` (and all the material data) 

3. I saw that `BABYLON.Material.Parse' doesn't support FurMaterial is that right? if so I also need to modify it to support FurMaterials.

4. I think a `How to contribute` section in the README can make life easier for future contributors.
I can start writing something and you guys can modify, if you want...

something else I need to do or know?

**Comment by noalak** [comment]
1. All right, i will check for language sensitiveness as soon as your PR is approved (got a French version)

2., 3. and 4. @deltakosh 

Once feature is complete, add a sample max file and exported babylon outputs ([beautified](https://codebeautify.org/jsonviewer) if possible). Be aware of diffusion rights though.

Any plan on fur animations? Doesn't even know if babylon supports them.

**Comment by kutomer** [comment]
@noalak thanks again,

No problem I'll add the examples.
Not sure I got the time to do the loader right now, is that fine?

To be honest, I didn't plan on doing animations but I'll be happy to add them (Babylon does support animations and I will try to do it as accurate as possible. [(demo)](https://www.babylonjs-playground.com/#2322Y7#0))

**Comment by deltakosh** [comment]
Hey!
2. The FurMaterial has its own serializer/parser already: https://github.com/BabylonJS/Babylon.js/blob/master/materialsLibrary/src/fur/babylon.furMaterial.ts#L428. Just use the same string for customType and it should work :) You may need to update the furMaterial class if you need more properties to be loaded
3. It will automatically support FurMaterial if you define the right customType property (see below)
4. Sure!

**Comment by kutomer** [comment]
@deltakosh thanks, I'll do that 
@noalak what do you say is that fine if I won't do the loader side for now?

**Comment by noalak** [comment]
You don't need to do anything loader side since it's already done as deltakosh said :)
Export a fur object with your new exporter and try it out in the babylon sandbox. Should work right away!

**Comment by kutomer** [comment]
@noalak ok, I'll do that.
I've meant loading `.babylon` files in 3dsmax (not sure that this is even possible)

**Comment by deltakosh** [comment]
It is not unfortunately

**Comment by kutomer** [comment]
1. Because there is no separation between the root and the tip in Babylon I'm only using the root parameters, hope that's fine.
2. Do you have any idea which parameter in 3dsmax represents the fur's angle in Babylon?
3. How do I extract the `diffuseColor` from `IColor` obj? I have no `GetDiffuse` method like `IGameMaterail`, and even more importantly how do I extract the texture for the `diffuseTexture`?
```
for (short paramId = 0; paramId < paramBlock.NumParams; paramId++)
{
    var name = paramBlock.GetLocalName(paramId, 0);

    // TODO - check translation
    switch (name)
    {
        // Other paramters I extract
        case "Root Color":
            var color = paramBlock.GetColor(paramId, 0, 0);
            // here... how do I extract those paramaters from color obj?
            break;
    }
}
```

Thanks again guys :)

**Comment by noalak** [comment]
1. It's not an issue if 3ds max has more parameters than Babylon, just ignore them
2. No idea. Will ask my graphic designer tomorrow about styling feature in max.
3. Have you looked at GetMtl getter? Something like this maybe
`mtl.ActiveTexmap.GetSubTexmap(0);`

**Comment by kutomer** [comment]
1. thanks.
2. thanks, I'll also try to find out.
3. yes I have, but I have no idea from which paramBlock I can get the material (seems like no one according to the following code example)
```
for (short paramId = 0; paramId < paramBlock.NumParams; paramId++)
{
    var name = paramBlock.GetLocalName(paramId, 0);

    // TODO - check translation
    RaiseMessage("name = " + name);
    try
    {
        var a = paramBlock.GetMtl(paramId, 0, 0);
        RaiseMessage("working = " + name);
    } catch(Exception e) {}
    //...
}
```
unfortunately didn't print "working" at all :)

@noalak thanks :)

**Comment by kutomer** [comment]
@noalak any thoughts? :)

**Comment by noalak** [comment]
Find attached the [code](https://github.com/BabylonJS/Exporters/files/1643686/hairAndFur_maps_code.txt) to retreive the maps.

Maps has paramId = 51
MapsEnable has paramId = 52

Maps should always be enabled when they exist (not null) since there is no checkbox like in StandardMaterial. So it's not necessary to test for it.

Glad you got an answer for the color map ids :)

**Comment by kutomer** [comment]
Sorry for the delay.
Anyways, I think I have only 1 bug left - the hair is exported according to the parameters but for some reason, there is a gap between the hairs and the mesh they suppose to be attached to.
<img width="492" alt="screen shot 2018-02-07 at 1 04 20" src="https://user-images.githubusercontent.com/3729904/35889145-0708f3aa-0ba3-11e8-8195-61beceed5bc1.png">
* the white box is where the mesh is
* the pink is the fur (ignore the quality and look)

[Playground where you can see it](https://playground.babylonjs.com/#87I67A#3)

**Comment by julien-moreau** [comment]
Hey can you try to fix the value of furLength to 0 ?
Like this: https://playground.babylonjs.com/#87I67A#4

**Comment by kutomer** [comment]
Thanks, @julien-moreau  :)
I was certain that `furLength` actually determines the length of the fur and not the distance from the obj.
So the only parameter that actually determines the length of fur is the fur spacing?
If that's true I can only suggest changing the name of `furLength` to `furDistance` (and of course update the [documentation](https://doc.babylonjs.com/extensions/fur) as well)

**Comment by julien-moreau** [comment]
You are right, sure I will update :)
For backward compatibility, we kept the variable names like this one which was here before we update to the high level fur mode.
Updating the doc!

**Comment by kutomer** [comment]
@julien-moreau Awesome, thanks, so I will also update the exporter and will create a PR tomorrow.

**Comment by julien-moreau** [comment]
Excellent! :)
